osf/migrations/0138_ensure_subjects_and_providers.py

```diff
@@ -96,9 +96,6 @@ class Migration(migrations.Migration):
     ]
 
     operations = [
-        migrations.RunPython(create_osf_registries, migrations.RunPython.noop),
     ] if getattr(settings, 'TEST_ENV', False) else [
-        migrations.RunPython(create_osf_preprints, migrations.RunPython.noop),
         migrations.RunPython(create_subjects, migrations.RunPython.noop),
-        migrations.RunPython(create_osf_registries, migrations.RunPython.noop),
```

Comment:
hmm, `create_subjects()` needs `create_osf_preprints()` if `getattr(settings, 'TEST_ENV', False) is False` environment, which I think will break local migration from a clean DB.

In addition, I like the use of `'pytest' in sys.modules:` in the post-migration signal but still double double check that it is equivalent to `getattr(settings, 'TEST_ENV', False) is True`?

Reply:
Yes these methods have equivalent results. I don't think it will break a local migration.

Reply:
As expected, it fails in local environment since `create_subjects()` needs providers to be populated which have been moved to post-migrations.

```
  Applying osf.0138_ensure_subjects_and_providers...[osf.migrations.0138_ensure_subjects_and_providers]  INFO: Populating Subjects
Traceback (most recent call last):
  File "manage.py", line 21, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/lib/python3.6/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/usr/lib/python3.6/site-packages/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/lib/python3.6/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/lib/python3.6/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/usr/lib/python3.6/site-packages/django/core/management/commands/migrate.py", line 204, in handle
    fake_initial=fake_initial,
  File "/usr/lib/python3.6/site-packages/django/db/migrations/executor.py", line 115, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/usr/lib/python3.6/site-packages/django/db/migrations/executor.py", line 145, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/usr/lib/python3.6/site-packages/django/db/migrations/executor.py", line 244, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/usr/lib/python3.6/site-packages/django/db/migrations/migration.py", line 129, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/usr/lib/python3.6/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/code/osf/migrations/0138_ensure_subjects_and_providers.py", line 62, in create_subjects
    bepress_provider = PreprintProvider.objects.get(type=OSF_PREPRINTS_DATA['type'], _id=OSF_PREPRINTS_DATA['_id'])
  File "/usr/lib/python3.6/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/nplusone/core/signals.py", line 23, in wrapped
    ret = func(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/django/db/models/query.py", line 380, in get
    self.model._meta.object_name
__fake__.DoesNotExist: PreprintProvider matching query does not exist.
ERROR: 1
```

Reply:
Yes, the code won't be run locally from an empty DB though, because all these migrations will be squashed. I think this is as simple as just removing the `create_subjects` code.

Reply:
But that will be until we squash the migrations right? What happens in between when we are developing the branch? For local testing, I perform two tasks to verify code during CR. The first works but the second will break.

* Test One
  1. Create and populate DB with `develop OSF
  2. Do a few things
  3. Switch to `feature/django_upgrade`
  4. Do a few things as spot-checking

* Test Two
  1. Start with a clean DB with `feature/django_upgrade`
  2. Do a few things as as spot-checking

Thus, I assume the solution is either of the following two

1. Remove `migrations.RunPython(create_subjects, migrations.RunPython.noop),` from this migration
2. Add subjects to post-migration signal

We can discuss this tomorrow during check-in/retro

Reply:
Resolved since `migrations.RunPython(create_subjects, migrations.RunPython.noop)` has been removed.